### PR TITLE
Modifications to Activation Extraction and Training of DCs

### PR DESCRIPTION
**Pull request by Kaleidophon**
Based on issue #6, the following changes were introduced:

- `Extractor` now takes an optional `selection_func` argument. The selection function determines whether the current activation should be extracted or not based on the current position, token and general sentence information.
- `DCTrainer` now automatically calculates the class weights for the training data and uses them to train the classifier.
- When creating the training and testing data for Diagnostic Classifiers, the whole data set is shuffled before splitting (before that, only the training data would be shuffled, the testing data remained the same)

### CODE REVIEW

File: rnnalyse/classifiers/dc_trainer.py [hunk 1]
```diff
@@ -65,8 +66,16 @@ def train(self, train_subset_size: int = -1, train_test_split: float = 0.9) -> N
         for a_name in self.activation_names:
             data_dict = self.activations_reader.create_data_split(a_name,
                                                                   train_subset_size,
+
```
Comment (dieuwkehupkes) [inline rnnalyse/classifiers/dc_trainer.py]: empty line?

File: rnnalyse/classifiers/dc_trainer.py [hunk 2]
```diff
@@ -65,8 +66,16 @@ def train(self, train_subset_size: int = -1, train_test_split: float = 0.9) -> N
...
                                                                   train_test_split)
 
+            # Calculate class weights
```
Comment (dieuwkehupkes) [inline rnnalyse/classifiers/dc_trainer.py]: Great! Should we maybe provide the option to overwrite this in case someone doesn't want to use class weights (for whatever reason)?

File: rnnalyse/extractors/base_extractor.py [hunk 3]
```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
+                    sen_activations[(l, name)] = np.append(
+                        sen_activations[(l, name)], activations[l][name].detach().numpy()[np.newaxis, ...], axis=0
+                    )
+
+        num_extracted = list(sen_activations.values())[0].shape[0]
+        self._dump_activations(sen_activations)
+
+        return num_extracted
+
+    def extract_average_eos_activations(self):
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: There is a lot of repetition here wrt to the `extract` function. If you'd move the selection function from init to the `extract` function, I think you could just reuse it to extract activations and then process them afterwards, right?

File: rnnalyse/extractors/base_extractor.py [hunk 4]
```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
         sentence : Sentence
             The to-be-extracted sentence, represented as a list of strings.
+
+        Returns
+        -------
+        num_extracted: int
+            Number of extracted activations for this sentence.
         """
+
+        sen_activations: PartialActivationDict = self._init_sen_activations(0)
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: This might make perfect sense, but I cannot quickly figure out why previously this was the sentence length and now it is 0, because you don't know on forehand if all activations are required?

File: rnnalyse/extractors/base_extractor.py [hunk 5]
```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
+
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: This seems sensible, but it is difficult to be sure if its correct. This would be an excellent functionality to write a test for that mocks the activation reader and tests if the selection actually writes away the correct activations.

@Kaleidophon could you create such a test? In machine you can find several examples of mocking.

File: rnnalyse/activations/activations_reader.py [hunk 6]
```diff
@@ -63,11 +64,13 @@ def create_data_split(self,
         split = int(self.data_len * train_test_split)
 
         n = split if train_subset_size == -1 else train_subset_size
-        train_indices = np.random.choice(range(n), n, replace=False)
+        indices = np.random.choice(range(self.data_len), self.data_len, replace=False)
```
[the remark this answers is not in the document]
Comment (dieuwkehupkes) [reply rnnalyse/activations/activations_reader.py]: It would actually be good to write a unittest for this.

File: rnnalyse/extractors/base_extractor.py [hunk 7]
```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
+        # Extract
+        for sentence in self.corpus.values():
+            sen_activations: PartialActivationDict = self._init_sen_activations(len(sentence))
 
-        for i, token in enumerate(sentence):
-            out, activations = self.model(token, activations)
+            activations: FullActivationDict = self.init_lstm_states.states
+
+            for i, token in enumerate(sentence.sen):
+                if self.selection_func(i, token, sentence):
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: Doesn't this skip words when the selection function doesn't select them? That would be problematic, because the activations are computed on non-sensical inputs then (imagine I have the sentence "I would like a coffee" and my selection function wants only the activations of nouns, then I would compute the activations by passing to the network "I coffee".

File: rnnalyse/extractors/base_extractor.py [hunk 8]
```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
 
-        # Update average eos states
-        self.avg_eos_states = {
-            l: {
-                'hx': _incremental_avg(self.avg_eos_states[l]['hx'], activations[l]['hx']),
-                'cx': _incremental_avg(self.avg_eos_states[l]['cx'], activations[l]['cx'])
+            # Update average eos states if last activation was extracted
+            avg_eos_states = {
+                l: {
+                    'hx': _incremental_avg(avg_eos_states[l]['hx'], activations[l]['hx']),
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: the reason that activations[l]['hx'] contains the last activation for the current sentence is that you don't store the intermediate ones?

File: rnnalyse/extractors/base_extractor.py [hunk 9]
```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
+        """ Extract average end of sentence activations and dump them to a file. """
+
         def _incremental_avg(old_avg, new_value):
             # n_sens only get incremented after _extract_sentence is called, therefore + 1 here
             return old_avg + 1 / (self.n_sens + 1) * (new_value - old_avg)
 
-        sen_activations: PartialActivationDict = self._init_sen_activations(len(sentence))
+        # Init states
+        avg_eos_states = {
+            l: {'hx': torch.zeros(self.model.hidden_size), 'cx': torch.zeros(self.model.hidden_size)}
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: It took me a while to figure out why we need the `l`. Maybe it would be more obviously clear if we renamed it `layer`?

File: rnnalyse/extractors/base_extractor.py [hunk 10]
```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
+                l: {
+                    'hx': _incremental_avg(avg_eos_states[l]['hx'], activations[l]['hx']),
+                    'cx': _incremental_avg(avg_eos_states[l]['cx'], activations[l]['cx'])
+                }
+                for l in avg_eos_states
             }
-            for l in self.avg_eos_states
-        }
 
-        self._dump_activations(sen_activations)
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: What is the reason you switched this line?

File: rnnalyse/activations/activations_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kaleidophon) [reply rnnalyse/activations/activations_reader.py]: How this behavior works could probably at least be specified in a docstring

File: rnnalyse/classifiers/dc_trainer.py (hunk 1 above)
Comment (Kaleidophon) [reply rnnalyse/classifiers/dc_trainer.py]: whoopsie

File: rnnalyse/classifiers/dc_trainer.py (hunk 2 above)
Comment (Kaleidophon) [reply rnnalyse/classifiers/dc_trainer.py]: Done

File: rnnalyse/extractors/base_extractor.py (hunk 3 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: Very good idea!

File: rnnalyse/extractors/base_extractor.py (hunk 4 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: Before, this was initialized as an empty np.array with dimensions sentence length x activation dim. Now that we have the selection function, we don't know in advance how many activations we will actually extract, that's why I opted to initialize it as 0 x activation dim instead and use `np.append`.

File: rnnalyse/extractors/base_extractor.py (hunk 5 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: See above. For testing I would just start opening and issue and collect the function we find especially worthy testing (or, differently put, worthy to prioritize for testing).

File: rnnalyse/extractors/base_extractor.py (hunk 7 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: Correct, that was a copy-paste error from an earlier, erroneous version. Corrected.

File: rnnalyse/extractors/base_extractor.py (hunk 8 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: I would say `activations[l]['hx']` contain the last activations because they're the last value that is assigned to this variable in the last loop of `for i, token in enumerate(sentence.sen)`. I actually just realized that I could omit the whole inner loop in this function but I am just going to reuse the code in `_extract_sentence` like you suggested.

File: rnnalyse/extractors/base_extractor.py (hunk 9 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: I was sticking to @jumelet's notation. As you know I like to write things out :-P

File: rnnalyse/extractors/base_extractor.py (hunk 10 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: Dump activations only dumps the "standard" activations. Now that I was putting this logic in it's own function, I needed to use a different dump function speficially for the eos activations

File: rnnalyse/activations/activations_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dieuwkehupkes) [reply rnnalyse/activations/activations_reader.py]: Yes there should be some documentation indeed. And probably change the name, if we don't even have the same idea about what it is supposed to do that's not good XD.
For me, having a functionality that lets you specify the exact amount of training points is not very useful, what I would find more useful is being able to not use the entire dataset that was is in the file. If you want to do a small test, for instance, you can then just specify that from the commandline, instead of having to create a new, smaller, file. @jumelet and @Kaleidophon, what do you think of changing the name of `train_subset_size` to `data_subset_size`, the interpretation is then a bit more intuitive (similar to 'cutoff' in the extract script).

File: rnnalyse/extractors/base_extractor.py (hunk 5 above)
Comment (dieuwkehupkes) [reply rnnalyse/extractors/base_extractor.py]: In general I agree that we should open issues for such things. But at the same time I believe that we are a bit too lacky about this. It is good practice to not accept PRs without some tests that show that they are correct, especially if they are function related (which may be difficult to see from a file). It is a bit strange perhaps to require that in a library that otherwise has no tests, but we have to start from somewhere.

@Kaleidophon  If you find this difficult, I can set something up and ask you to finish the actual implementation of the test logic?

File: rnnalyse/extractors/base_extractor.py (hunk 9 above)
Comment (dieuwkehupkes) [reply rnnalyse/extractors/base_extractor.py]: Hahaha I was already surprised indeed. I like to optimise making things understandable and keeping lines short. E.g., I don't mind using abbreviations in certain contexts, such as act_len instead of activation_length, or use single characters for small loops for which the meaning is not that important, but I'm also for helping the code reader as much as possible.

I'm sure we can figure out some style that makes us all happy :)

File: rnnalyse/extractors/base_extractor.py (hunk 5 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: I can do that. If you provide some form of template that would be useful to me in order to see what you expect

File: rnnalyse/activations/activations_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kaleidophon) [reply rnnalyse/activations/activations_reader.py]: Mh, tbh I don't have an opinion of what would be more useful. @jumelet ?

[the remark this answers is not in the document]
Comment (dieuwkehupkes) [reply rnnalyse/activations/activations_reader.py]: Ak then question: in the current formulation, aren't they just the same formulations of a different thing? If you define `train_subset size` this decides already the split, and if you define `train_test_split' this determines the subset size of the training data. 

Of course, sometimes it can be very useful to have two ways to specify the same thing because one specification may be more easily accessible, but in this case I don't really see that happening.
@Kaleidophon if you find these two different ways of specifying how much of the data to use for training useful, I'm curious what kind of situations do you think it would be more convenient to use `train_subset_size` instead of the split. Maybe I'm not considering them.

File: rnnalyse/extractors/base_extractor.py [hunk 11]
```diff
@@ -147,38 +146,73 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
...
+        # Init states
+        avg_eos_states = {
+            layer: {'hx': torch.zeros(self.model.hidden_size), 'cx': torch.zeros(self.model.hidden_size)}
+            for layer in range(self.model.num_layers)
         }
 
-        self._dump_activations(sen_activations)
+        # Extract
+        for i, labeled_sentence in enumerate(self.corpus.values()):
+            sen_activations = self._extract_sentence(labeled_sentence, _eos_selection_func)
```
Comment (dieuwkehupkes) [inline rnnalyse/extractors/base_extractor.py]: These are already the eos activations, right? not the sentence activations? Maybe reflect that in the name of the variable (eos_activations = ...)?

File: test/test_basic_extractor.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+from test_utils import create_sentence_dummy_activations
+
+
+# GLOBALS
+ACTIVATION_DIM = 10
+ACTIVATION_NAMES = [(0, "hx"), (0, "cx")]
+ACTIVATIONS_DIR = "test_data"
+
+
+class MockLanguageModel(LanguageModel):
```
Comment (dieuwkehupkes) [inline test/test_basic_extractor.py]: I think this could also be directly mocked by decorating the function where it is called with mock.patch or mock.patch.object. This is also a nice solution though.

File: test/test_basic_extractor.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+    def test_extract_sentence(self):
+        """ Test the _extract_sentence class, especially with focus on the selection function functionality. """
+
+        def _merge_sentence_activations(sentences_activations):
+            return np.array(list(itertools.chain(
+                *[sentence_activations[(0, "hx")] for sentence_activations in sentences_activations])
+            ))
+
+        # Test extraction of all activations
```
Comment (dieuwkehupkes) [inline test/test_basic_extractor.py]: maybe it makes sense to put these different ones in different functions (test_extract_by_token, test_extract_by_position... etc)?

File: test/test_basic_extractor.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        assert token == next_token
+
+        return None, {0: {"hx": next_activation, "cx": next_activation}}
+
+    def reset(self):
+        """ Reset the activations for next test. """
+        self.all_pairs = zip(self.all_tokens, self.all_activations)
+
+
+class TestExtractor(unittest.TestCase):
```
Comment (dieuwkehupkes) [inline test/test_basic_extractor.py]: maybe we can also add a test that tests if the pickle functions are called with the right arguments?

File: test/test_dc_trainer.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+    @patch('rnnalyse.activations.activations_reader.ActivationsReader.create_data_split')
+    @patch('rnnalyse.classifiers.dc_trainer.DCTrainer._reset_classifier')
+    @patch('rnnalyse.classifiers.dc_trainer.DCTrainer.log_results')
+    @patch('rnnalyse.classifiers.dc_trainer.DCTrainer.save_classifier')
+    @patch('rnnalyse.classifiers.dc_trainer.DCTrainer.eval_classifier')
+    @patch('rnnalyse.classifiers.dc_trainer.DCTrainer.fit_data')
+    def test_class_weights(self, mock_fit_data, mock_eval_classifier, mock_save_classifier, mock_log_results,
+                           mock_reset_classifier, create_data_split_mock):
+        create_data_split_mock.return_value = self.data_dict
```
Comment (dieuwkehupkes) [inline test/test_dc_trainer.py]: FYI I think it is possible to do this also directly in the patch by specifying `return_value = self.data_dict`?

File: rnnalyse/activations/activations_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kaleidophon) [reply rnnalyse/activations/activations_reader.py]: I just see them as a relative and an absolute way to specify the split. I could see that someone might want to say "I want to use _exactly_ 1000 samples to train my model" but I also acknowledge that usually it's done by giving a percentage. That being said, I don't mind leaving it there just in case someone needs it but I am also not attached to it and open to remove or replace it.

File: rnnalyse/extractors/base_extractor.py (hunk 11 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: Done!

File: test/test_activations_reader.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        )
+        self.num_labels = labels.shape[0]
+        self.activation_reader = ActivationsReader(activations_dir=ACTIVATIONS_DIR)
+
+    def test_read_activations(self):
+        activations = self.activation_reader.read_activations((0, "hx"))
+        labels = self.activation_reader.labels
+
+        # Check if the amount of read data is correct
+        assert self.num_labels == labels.shape[0] == activations.shape[0]
```
[the remark this answers is not in the document]
Comment (Kaleidophon) [reply test/test_activations_reader.py]: Done!

File: test/test_basic_extractor.py (hunk 12 above)
Comment (Kaleidophon) [reply test/test_basic_extractor.py]: The way I have the model produce these dummy activations in an ordered manner makes this unfortunately very tricky (when mocking the forward function, I need access to `self.all_pairs`. After one test when this list was consumed by `next(...)` I want to reset it for the next test, but when using Mocks I have no (easy) way (I found) to access this variable again). Therefore I opted for a custom mock class in this case.

File: test/test_dc_trainer.py (hunk 15 above)
Comment (Kaleidophon) [reply test/test_dc_trainer.py]: Usually yes, but I don't have access to `self` inside the decorator :-(

Comment (dieuwkehupkes) [reply test/test_dc_trainer.py]: Ah I see, so you'd have to define it somewhere else. I can see how that's not optimal.

File: rnnalyse/activations/activations_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dieuwkehupkes) [reply rnnalyse/activations/activations_reader.py]: Then I propose to do it as I suggested before, and use the argument `data_subset_size` as a  way for people to only use part of their dataset, similar to the `cutoff` argument in the extractor

File: test/test_basic_extractor.py (hunk 14 above)
Comment (Kaleidophon) [reply test/test_basic_extractor.py]: Done!

File: test/test_basic_extractor.py (hunk 13 above)
Comment (Kaleidophon) [reply test/test_basic_extractor.py]: Done!

File: rnnalyse/activations/activations_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kaleidophon) [reply rnnalyse/activations/activations_reader.py]: Done (with tests)!
